### PR TITLE
 22Q2-7 ✅ Add upgrade for moving NS ownership to cluster

### PR DESCRIPTION
**Pull request by deifyed**
## Description
- Adds an ArgoCD application named namespaces tracking the directory
    `infrastructure/<cluster>/argocd/namespaces`
- Creates new namespace.yaml files in
    `infrastructure/<cluster>/argocd/namespaces` based on the namespace
	manifests found in `infrastructure/applications/<app/base` folders
- Deletes old `infrastructure/applications/<app>/base/namespace.yaml` files iff
    all clusters who has the application running has migrated to the new style

## Motivation and Context

[22Q2-7](https://trello.com/c/5nKN6net)

### A single cluster setup
1. Run the upgrade
2. Verify that no `namespace.yaml` files are left in
    `infrastructure/applications/*/base`
3. Verify that the relevant namespaces are represented in
    `infrastructure/<cluster/argocd/namespaces`

### Multicluster setup
1. Run the upgrade

Depending on the application and the setup, do one of the following
- Verify that the only namespaces that have been removed are in applications
    that only belong to one cluster
- Verify that the only namespaces that have been removed are in applications
    where all relevant clusters has been upgraded

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Review by yngvark — APPROVED** [review]
Can't find any bugs, looks good.

However, it is quite tough to try to verify the logic and trying to find any bugs, it would require lots of time. A better approach is good testing. So, to avoid potential problems later when we have started forgotting how this upgrade works, I would recommend you to test this upgrade now by running it on a copy of each of our teams' IAC repositories. Especially if this upgrade might affect production clusters when running it in dev.

File: upgrades/0.0.99.seperate-ns-from-app/pkg/migration/migrations_test.go [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+	appBaseDir := path.Join("/infrastructure/applications", appName, "base")
+
+	err := fs.MkdirAll(appBaseDir, paths.DefaultFolderPermissions)
+	assert.NoError(t, err)
+
+	err = fs.WriteReader(path.Join(appBaseDir, "namespace.yaml"), namespace(namespaceName))
+	assert.NoError(t, err)
+}
+
+func TestMigrateApplication(t *testing.T) {
```
Comment (yngvark) [inline upgrades/0.0.99.seperate-ns-from-app/pkg/migration/migrations_test.go:111]: I didn't think of this before today, but anyway, on possible change that might improve readability is this:

From the outside, what I really care about when running this upgrade, is to understand where were my files before, and where are they now. That is how I think I would be able to understand what happens best.

The test code could reflect this by doing somethng like this:

```golang
	testCases := []struct {
		// ...
		withFiles []string
		expectedFilesAfter []string
	}{
		{
			withFiles: []string{
				// Base folder
				// - Hello app
				"infrastructure/applications/hello/base/namespace.yaml",
				// Perhaps add more files to make the example realistic? Or perhaps it is just noise.

				// - Some other app
				"infrastructure/applications/some-other-app/base/namespace.yaml",

				// Cluster folder
				"infrastructure/my-cluster/argocd/applications/applications.yaml",

				// ... more files
			},
			expectedFilesAfter: []string{
				// Cluster folder
				"infrastructure/my-cluster/argocd/applications/namespaces.yaml",
				"infrastructure/my-cluster/argocd/namespaces/hello.yaml",
				"infrastructure/my-cluster/argocd/namespaces/some-other-namespace.yaml",
			},
		}
	}
```

Comment (deifyed) [reply upgrades/0.0.99.seperate-ns-from-app/pkg/migration/migrations_test.go:113]: https://github.com/oslokommune/okctl-upgrade/pull/25/commits/5c5212cb985c93fcc1f0bd1ccb282cb4e37113d6

Comment (deifyed) [reply upgrades/0.0.99.seperate-ns-from-app/pkg/migration/migrations_test.go:113]: Nice suggestion

File: upgrades/0.0.99.seperate-ns-from-app/makefile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+GOLANGCILINT   := $(GOBIN)/golangci-lint
+GOFUMPT := $(GOBIN)/gofumpt
+
+$(RICHGO):
+	$(GO) install github.com/kyoh86/richgo@v0.3.6
+
+$(GOLANGCILINT):
+	# To bump, simply change the version at the end to the desired version. The git sha here points to the newest       commit
+	# of the install script verified by our team located here: https://github.com/golangci/golangci-lint/blob/mast      er/install.sh
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/b90551cdf9c6214075f2a40d1b5595c6b41ffff0/i      nstall.sh | sh -s -- -b ${GOBIN} v1.43.0
```
[the remark this answers is not in the document]
Comment (deifyed) [reply upgrades/0.0.99.seperate-ns-from-app/makefile:15]: https://c.tenor.com/84DHgXJ8QAoAAAAC/tenor.gif

[the remark this answers is not in the document]
Comment (deifyed) [reply upgrades/0.0.99.seperate-ns-from-app/makefile]: https://github.com/oslokommune/okctl-upgrade/pull/25/commits/12cdc9c5eb373d6e059400e5e5222c5a7f3be014

**Pull request merged**